# An introduction to the Pandas package 
This is a companion notebook for video content presented as part of the Data Mining course at SMU.

In this tutorial we will be looking at a number of different parts of the Pandas package for data analysis, including:
- Data Frames
 - loading data
 - head and tail commands
- Munging
 - indexing operations
 - basic statistics
 - encoding
 - imputation (optional, but recommended)


## Data Frames
Data frames in Pandas are basically like tables of data that you can do some really interesting relational database operations upon. There are many built in methods for aggregation and visualization, but we will cover those next time.

First lets load a typical table of data from a csv file. You can download the file from here:
https://archive.ics.uci.edu/ml/datasets/Heart+Disease

Make sure to place it in this directory or adjust the path for the file.
### Reading Data from CSV with Pandas

In [ ]:
# let's print out the first five rows inside a csv file

# NOTE: you may need to change the path to the file, 
#       depending on where you saved the data
with open('data/heart_disease.csv') as fid:
    for idx, row in enumerate(fid):
        print(row,end='')
        if idx >= 4:
            break

In [ ]:
# now let's read in the same data using pandas to save it as a dataframe
import pandas as pd

df = pd.read_csv('data/heart_disease.csv') # read in the csv file

In [4]:
# now lets look at the data
df.head()

site  age  is_male  chest_pain rest_blood_press cholesterol  \
0  cleve   63        1           1              145         233   
1  cleve   67        1           4              160         286   
2  cleve   67        1           4              120         229   
3  cleve   37        1           3              130         250   
4  cleve   41        0           2              130         204   

  high_blood_sugar rest_ecg max_heart_rate exer_angina ST_depression  \
0                1        2            150           0           2.3   
1                0        2            108           1           1.5   
2                0        2            129           1           2.6   
3                0        0            187           0           3.5   
4                0        2            172           0           1.4   

  Peak_ST_seg major_vessels thal  has_heart_disease  
0           3             0    6                  0  
1           2             3    3                  2  
2           2             2    7                  1  
3           3             0    3                  0  
4           1             0    3                  0

In [5]:
# now let's a get a summary of the variables 
print(df.info())
# we can see that most of the data 
#  is saved as an integer or as a nominal object

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 15 columns):
site                 920 non-null object
age                  920 non-null int64
is_male              920 non-null int64
chest_pain           920 non-null int64
rest_blood_press     920 non-null object
cholesterol          920 non-null object
high_blood_sugar     920 non-null object
rest_ecg             920 non-null object
max_heart_rate       920 non-null object
exer_angina          920 non-null object
ST_depression        920 non-null object
Peak_ST_seg          920 non-null object
major_vessels        920 non-null object
thal                 920 non-null object
has_heart_disease    920 non-null int64
dtypes: int64(4), object(11)
memory usage: 107.9+ KB
None


This data has been read into working memory and is known as a DataFrame.

### Reading Data from SQLite3 with Pandas
We can also connect to a sqlite3 database using the built in sqlite3 package that ships with python. 

In [6]:
# but csv files are not the only thing we can work with
# what if the data was actually in a sqlite database?
del df
import sqlite3

con = sqlite3.connect('data/heart_disease_sql') # again this file is in the same directory
df = pd.read_sql('SELECT * FROM heart_disease', con)  # the table name is heart_disease
df.head()

site age is_male chest_pain rest_blood_press cholesterol high_blood_sugar  \
0  cleve  63       1          1              145         233                1   
1  cleve  67       1          4              160         286                0   
2  cleve  67       1          4              120         229                0   
3  cleve  37       1          3              130         250                0   
4  cleve  41       0          2              130         204                0   

  rest_ecg max_heart_rate exer_angina ST_depression Peak_ST_seg major_vessels  \
0        2            150           0           2.3           3             0   
1        2            108           1           1.5           2             3   
2        2            129           1           2.6           2             2   
3        0            187           0           3.5           3             0   
4        2            172           0           1.4           1             0   

  thal has_heart_disease  
0    6                 0  
1    3                 2  
2    7                 1  
3    3                 0  
4    3                 0

In [7]:
df.info()
# notice now, however, that the data types are all objects!

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 15 columns):
site                 920 non-null object
age                  920 non-null object
is_male              920 non-null object
chest_pain           920 non-null object
rest_blood_press     920 non-null object
cholesterol          920 non-null object
high_blood_sugar     920 non-null object
rest_ecg             920 non-null object
max_heart_rate       920 non-null object
exer_angina          920 non-null object
ST_depression        920 non-null object
Peak_ST_seg          920 non-null object
major_vessels        920 non-null object
thal                 920 non-null object
has_heart_disease    920 non-null object
dtypes: object(15)
memory usage: 107.9+ KB


 ### Working with DataFrames
 We can index in to DataFrame in a number of ways:

In [8]:
# the variable names are embedded into the structure
print (df.age)
print (df['age']) # but can also be accessed using strings

0      63
1      67
2      67
3      37
4      41
5      56
6      62
7      57
8      63
9      53
10     57
11     56
12     56
13     44
14     52
15     57
16     48
17     54
18     48
19     49
20     64
21     58
22     58
23     58
24     60
25     50
26     58
27     66
28     43
29     40
       ..
890    52
891    53
892    53
893    54
894    55
895    55
896    55
897    56
898    56
899    56
900    58
901    59
902    59
903    65
904    66
905    41
906    43
907    44
908    47
909    47
910    49
911    49
912    50
913    50
914    52
915    52
916    54
917    56
918    58
919    65
Name: age, dtype: object
0      63
1      67
2      67
3      37
4      41
5      56
6      62
7      57
8      63
9      53
10     57
11     56
12     56
13     44
14     52
15     57
16     48
17     54
18     48
19     49
20     64
21     58
22     58
23     58
24     60
25     50
26     58
27     66
28     43
29     40
       ..
890    52
891    53
892    53
893    54
894    55
895  

In [9]:
print (df.chest_pain.min(), df.chest_pain.max(), df.chest_pain.mean())

1 4 inf


In [10]:
# lets get rid of the 'site' variable
if 'site' in df:
    del df['site']

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null object
is_male              920 non-null object
chest_pain           920 non-null object
rest_blood_press     920 non-null object
cholesterol          920 non-null object
high_blood_sugar     920 non-null object
rest_ecg             920 non-null object
max_heart_rate       920 non-null object
exer_angina          920 non-null object
ST_depression        920 non-null object
Peak_ST_seg          920 non-null object
major_vessels        920 non-null object
thal                 920 non-null object
has_heart_disease    920 non-null object
dtypes: object(14)
memory usage: 100.7+ KB


In [11]:
# Notice that all of the data is stored as a non-null object
# That's not good. It means we need to change those data types
# in order to encode the variables properly. Right now Pandas
# thinks all of our variables are nominal!

import numpy as np
# replace '?' with -1, we will deal with missing values later
df = df.replace(to_replace='?',value=-1) # don't set to NaN, becasue that's an error if data isn't float

# let's start by first changing the numeric values to be floats
continuous_features = ['rest_blood_press', 'cholesterol', 
                       'max_heart_rate', 'ST_depression']

# and the oridnal values to be integers
ordinal_features = ['age','major_vessels','chest_pain',
                    'rest_ecg','Peak_ST_seg','thal','has_heart_disease']

# we won't touch these variables, keep them as categorical
categ_features = ['is_male','high_blood_sugar','exer_angina'];

# use the "astype" function to change the variable type
df[continuous_features] = df[continuous_features].astype(np.float64)
df[ordinal_features] = df[ordinal_features].astype(np.int64)

df.info() # now our data looks better!!

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null int64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null int64
major_vessels        920 non-null int64
thal                 920 non-null int64
has_heart_disease    920 non-null int64
dtypes: float64(4), int64(7), object(3)
memory usage: 100.7+ KB


In [12]:
df.head()

age is_male  chest_pain  rest_blood_press  cholesterol high_blood_sugar  \
0   63       1           1             145.0        233.0                1   
1   67       1           4             160.0        286.0                0   
2   67       1           4             120.0        229.0                0   
3   37       1           3             130.0        250.0                0   
4   41       0           2             130.0        204.0                0   

   rest_ecg  max_heart_rate exer_angina  ST_depression  Peak_ST_seg  \
0         2           150.0           0            2.3            3   
1         2           108.0           1            1.5            2   
2         2           129.0           1            2.6            2   
3         0           187.0           0            3.5            3   
4         2           172.0           0            1.4            1   

   major_vessels  thal  has_heart_disease  
0              0     6                  0  
1              3     3                  2  
2              2     7                  1  
3              0     3                  0  
4              0     3                  0

Let's get summary of all attributes in the frame

In [13]:
df.describe() # will get summary of continuous or the nominals

age  chest_pain  rest_blood_press  cholesterol    rest_ecg  \
count  920.000000  920.000000        920.000000   920.000000  920.000000   
mean    53.510870    3.250000        123.594565   192.604348    0.601087   
std      9.424685    0.930969         37.484705   114.615011    0.808415   
min     28.000000    1.000000         -1.000000    -1.000000   -1.000000   
25%     47.000000    3.000000        120.000000   164.000000    0.000000   
50%     54.000000    4.000000        130.000000   221.000000    0.000000   
75%     60.000000    4.000000        140.000000   267.000000    1.000000   
max     77.000000    4.000000        200.000000   603.000000    2.000000   

       max_heart_rate  ST_depression  Peak_ST_seg  major_vessels        thal  \
count      920.000000     920.000000   920.000000     920.000000  920.000000   
mean       129.263043       0.752174     0.840217      -0.436957    1.871739   
std         41.376773       1.154353     1.403211       0.959656    3.313649   
min         -1.000000      -2.600000    -1.000000      -1.000000   -1.000000   
25%        115.000000       0.000000    -1.000000      -1.000000   -1.000000   
50%        138.000000       0.200000     1.000000      -1.000000   -1.000000   
75%        156.000000       1.500000     2.000000       0.000000    6.000000   
max        202.000000       6.200000     3.000000       3.000000    7.000000   

       has_heart_disease  
count         920.000000  
mean            0.995652  
std             1.142693  
min             0.000000  
25%             0.000000  
50%             1.000000  
75%             2.000000  
max             4.000000

There are 920 entries in this data frame. Notice that this data frame has a number of missing values denoted by the value -1 (that we changed the '?' value to before). We need to either remove the missing values from the dataset OR we need to fill in with our best guess for those values. Let's first drop all the rows with missing values.

In [14]:
# how many value have the -1 (which we set as the missing values) 
import numpy as np

# let's set those values to NaN, so that Pandas understand they are missing
df = df.replace(to_replace=-1,value=np.nan) # replace -1 with NaN (not a number)
print (df.info())
df.describe() # scroll over to see the values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     861 non-null float64
cholesterol          890 non-null float64
high_blood_sugar     830 non-null object
rest_ecg             918 non-null float64
max_heart_rate       865 non-null float64
exer_angina          865 non-null object
ST_depression        856 non-null float64
Peak_ST_seg          611 non-null float64
major_vessels        309 non-null float64
thal                 434 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3), object(3)
memory usage: 100.7+ KB
None


/Users/eclarson/anaconda/envs/MLEnv/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


age  chest_pain  rest_blood_press  cholesterol    rest_ecg  \
count  920.000000  920.000000        861.000000   890.000000  918.000000   
mean    53.510870    3.250000        132.132404   199.130337    0.604575   
std      9.424685    0.930969         19.066070   110.780810    0.805827   
min     28.000000    1.000000          0.000000     0.000000    0.000000   
25%     47.000000    3.000000               NaN          NaN         NaN   
50%     54.000000    4.000000               NaN          NaN         NaN   
75%     60.000000    4.000000               NaN          NaN         NaN   
max     77.000000    4.000000        200.000000   603.000000    2.000000   

       max_heart_rate  ST_depression  Peak_ST_seg  major_vessels        thal  \
count      865.000000     856.000000   611.000000     309.000000  434.000000   
mean       137.545665       0.883178     1.770867       0.676375    5.087558   
std         25.926276       1.088707     0.619256       0.935653    1.919075   
min         60.000000      -2.600000     1.000000       0.000000    3.000000   
25%               NaN            NaN          NaN            NaN         NaN   
50%               NaN            NaN          NaN            NaN         NaN   
75%               NaN            NaN          NaN            NaN         NaN   
max        202.000000       6.200000     3.000000       3.000000    7.000000   

       has_heart_disease  
count         920.000000  
mean            0.995652  
std             1.142693  
min             0.000000  
25%             0.000000  
50%             1.000000  
75%             2.000000  
max             4.000000

Wow. Notice how the number of attributes went down in the description function. Looks like we need to impute values. If we drop the rows with missing data, we will be throwing away almost 80% of the data collected. No way!!

### Imputation of NaN values (Optional)

In [15]:
# lets look at some stats of the data
df.median() # only calculates for numeric data

age                   54.0
is_male                1.0
chest_pain             4.0
rest_blood_press     130.0
cholesterol          223.0
high_blood_sugar       0.0
rest_ecg               0.0
max_heart_rate       140.0
exer_angina            0.0
ST_depression          0.5
Peak_ST_seg            2.0
major_vessels          0.0
thal                   6.0
has_heart_disease      1.0
dtype: float64

In [16]:
# the 'fillna' function will take the given series (the output above)
# and fill in the missing values for the columns it has
df_imputed = df.fillna(df.median()) # note that to do this all values must be numeric
df_imputed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null float64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null float64
major_vessels        920 non-null float64
thal                 920 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3), object(3)
memory usage: 100.7+ KB


Notice that the object variables are unchanged, but all the numeric/ordinal values have been filled in with the median of the columns. Let's try something (slightly) smarter, and fill in the oridinals with the median and the continuous with the mean.

In [17]:
# make  one series for imputing with
series_mean = df[continuous_features].mean()
series_median = df[categ_features+ordinal_features].median()
cat_series = pd.concat((series_median,series_mean))

print (cat_series)

is_male                1.000000
high_blood_sugar       0.000000
exer_angina            0.000000
age                   54.000000
major_vessels          0.000000
chest_pain             4.000000
rest_ecg               0.000000
Peak_ST_seg            2.000000
thal                   6.000000
has_heart_disease      1.000000
rest_blood_press     132.132404
cholesterol          199.130337
max_heart_rate       137.545665
ST_depression          0.883178
dtype: float64


In [18]:
# now let's impute the numbers a bit differently

df_imputed = df.fillna(value=cat_series)
df_imputed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null float64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null float64
major_vessels        920 non-null float64
thal                 920 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3), object(3)
memory usage: 100.7+ KB


In [28]:
df_imputed[categ_features].describe()

is_male high_blood_sugar exer_angina
count      920              830         865
unique       2                2           2
top          1                0           0
freq       726              692         528

In [29]:
# what if we want to impute values after grouping? Split-Impute-Combine
df_grouped = df.groupby(by=['is_male','high_blood_sugar'])

In [30]:
# now use this grouping to fill the data set in each group, then transform back
df_imputed = df_grouped.transform(lambda grp: grp.fillna(grp.mean()))
print(df_imputed.info())
print('------------')

# the above process can remove columns, so let's find that and fix it
names_removed = df.columns - df_imputed.columns # use list like a set operator
print(names_removed)
print('------------')

df_imputed[names_removed] = df[names_removed]
print(df_imputed.info())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 11 columns):
age                  920 non-null int64
chest_pain           920 non-null int64
rest_blood_press     919 non-null float64
cholesterol          920 non-null float64
rest_ecg             919 non-null float64
max_heart_rate       920 non-null float64
ST_depression        915 non-null float64
Peak_ST_seg          899 non-null float64
major_vessels        835 non-null float64
thal                 858 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3)
memory usage: 79.1 KB
None
------------
Index(['exer_angina', 'high_blood_sugar', 'is_male'], dtype='object')
------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
chest_pain           920 non-null int64
rest_blood_press     919 non-null float64
cholesterol          920 non-null float64
rest_ecg          

/Users/eclarson/anaconda/envs/MLEnv/lib/python3.5/site-packages/ipykernel/__main__.py:7: FutureWarning: using '-' to provide set differences with Indexes is deprecated, use .difference()


### Indexing logically into Data Frames
Let's now say that we are only interested in the summary of the dataframe when the patient has heart disease. We can achieve this using a few line of code:

In [31]:
df_imputed[df_imputed.has_heart_disease==0].describe()

/Users/eclarson/anaconda/envs/MLEnv/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


age  chest_pain  rest_blood_press  cholesterol    rest_ecg  \
count  411.000000  411.000000        411.000000   411.000000  411.000000   
mean    50.547445    2.761557        130.075507   227.713509    0.547445   
std      9.433700    0.903425         16.486825    74.217690    0.805204   
min     28.000000    1.000000         80.000000     0.000000    0.000000   
25%     43.000000    2.000000        120.000000   201.000000    0.000000   
50%     51.000000    3.000000        130.000000   227.000000    0.000000   
75%     57.000000    4.000000        140.000000   266.500000    1.000000   
max     76.000000    4.000000        190.000000   564.000000    2.000000   

       max_heart_rate  ST_depression  Peak_ST_seg  major_vessels        thal  \
count      411.000000     411.000000   405.000000     398.000000  398.000000   
mean       148.235344       0.445843     1.610641       0.491867    4.486785   
std         23.175084       0.707479     0.474372       0.462333    1.368367   
min         69.000000      -1.100000     1.000000       0.000000    3.000000   
25%        135.000000       0.000000          NaN            NaN         NaN   
50%        150.000000       0.000000          NaN            NaN         NaN   
75%        165.000000       0.938866          NaN            NaN         NaN   
max        202.000000       4.200000     3.000000       3.000000    7.000000   

       has_heart_disease  
count              411.0  
mean                 0.0  
std                  0.0  
min                  0.0  
25%                  0.0  
50%                  0.0  
75%                  0.0  
max                  0.0

In [32]:
# or we can use the extremely useful "groupby" function
df_imputed.groupby(by='has_heart_disease').median()

age  chest_pain  rest_blood_press  cholesterol  rest_ecg  \
has_heart_disease                                                             
0                   51           3             130.0        227.0       0.0   
1                   55           4             130.0        226.0       0.0   
2                   58           4             134.0        193.0       0.0   
3                   60           4             132.0        212.0       1.0   
4                   59           4             137.0        218.5       1.0   

                   max_heart_rate  ST_depression  Peak_ST_seg  major_vessels  \
has_heart_disease                                                              
0                           150.0           0.00     1.782991       0.674286   
1                           130.0           1.00     2.000000       0.674286   
2                           130.0           1.40     2.000000       0.674286   
3                           122.0           1.20     2.000000       0.805893   
4                           126.5           2.45     2.000000       0.937500   

                       thal  
has_heart_disease            
0                  3.734043  
1                  5.428571  
2                  6.000000  
3                  5.759259  
4                  6.000000

In [33]:
df_imputed.groupby(by=df_imputed.has_heart_disease>0).mean()

age  chest_pain  rest_blood_press  cholesterol  \
has_heart_disease                                                         
False              50.547445    2.761557        130.075507   227.713509   
True               55.903733    3.644401        134.100951   177.191037   

                   rest_ecg  max_heart_rate  ST_depression  Peak_ST_seg  \
has_heart_disease                                                         
False              0.547445      148.235344       0.445843     1.610641   
True               0.650749      128.757144       1.251465     1.908871   

                   major_vessels      thal  has_heart_disease  
has_heart_disease                                              
False                   0.491867  4.486785           0.000000  
True                    0.863240  5.697933           1.799607

In [34]:
df_imputed.groupby(by=df_imputed.major_vessels>2).mean()

age  chest_pain  rest_blood_press  cholesterol  rest_ecg  \
major_vessels                                                                   
False          53.368889    3.241111        132.198349   198.301656  0.587965   
True           59.900000    3.650000        136.900000   265.450000  1.350000   

               max_heart_rate  ST_depression  Peak_ST_seg  major_vessels  \
major_vessels                                                              
False              137.523459       0.867799     1.772801       0.629447   
True               134.550000       1.865000     1.850000       3.000000   

                   thal  has_heart_disease  
major_vessels                               
False          5.122661           0.966667  
True           5.700000           2.300000

### One Hot Encoding of Categorical Variables

In [35]:
# one hot encoded variables can be created using the get_dummies variable
tmpdf = pd.get_dummies(df_imputed['chest_pain'],prefix='chest')

tmpdf.head()

chest_1  chest_2  chest_3  chest_4
0      1.0      0.0      0.0      0.0
1      0.0      0.0      0.0      1.0
2      0.0      0.0      0.0      1.0
3      0.0      0.0      1.0      0.0
4      0.0      1.0      0.0      0.0

In [36]:
#one hot encoding of ALL categorical variables
# there is lot going on in this one line of code, so let's step through it

# pd.concat([*]], axis=1) // this line of code concatenates all the data frames in the [*] list
# [** for col in categ_features] // this steps through each feature in categ_features and 
#                                //   creates a new element in a list based on the output of **
# pd.get_dummies(df_imputed[col],prefix=col) // this creates a one hot encoded dataframe of the variable=col (like code above)

one_hot_df = pd.concat([pd.get_dummies(df_imputed[col],prefix=col) for col in categ_features], axis=1)

one_hot_df.head()

is_male_0  is_male_1  high_blood_sugar_0  high_blood_sugar_1  \
0        0.0        1.0                 0.0                 1.0   
1        0.0        1.0                 1.0                 0.0   
2        0.0        1.0                 1.0                 0.0   
3        0.0        1.0                 1.0                 0.0   
4        1.0        0.0                 1.0                 0.0   

   exer_angina_0  exer_angina_1  
0            1.0            0.0  
1            0.0            1.0  
2            0.0            1.0  
3            1.0            0.0  
4            1.0            0.0